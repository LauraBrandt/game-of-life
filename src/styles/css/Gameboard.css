.gameboard {
  border: 3px solid rgba(0, 0, 0, 0.5);
  background-color: #fafafa;
  margin: 0 1rem; }

.board-row {
  margin: 0;
  padding: 0;
  font-size: 0;
  display: flex; }

.board-cell {
  height: 16px;
  width: 16px;
  display: inline-block;
  border-right: 1px solid #aaa;
  border-bottom: 1px solid #aaa; }
  @media screen and (max-width: 850px) {
    .board-cell {
      height: 14px;
      width: 14px; } }
  @media screen and (max-width: 750px) {
    .board-cell {
      height: 12px;
      width: 12px; } }
  @media screen and (max-width: 600px) {
    .board-cell {
      height: 9px;
      width: 9px; } }
  .board-cell.active {
    background-color: #222; }
