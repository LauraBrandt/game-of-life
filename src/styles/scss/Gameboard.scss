@import 'variables';

.gameboard {
  border: 3px solid rgba(0,0,0,0.5);
  background-color: $light;
  margin: 0 1rem;
}

.board-row {
  margin: 0;
  padding: 0;
  font-size: 0;
  display: flex;
}

.board-cell {
  height: $cell-size;
  width: $cell-size;
  display: inline-block;
  border-right: 1px solid $grey;
  border-bottom: 1px solid $grey;
  @media screen and (max-width: $break-md-sm) {
    height: $cell-size-md;
    width: $cell-size-md;
  }
  @media screen and (max-width: $break-sm) {
    height: $cell-size-sm;
    width: $cell-size-sm;
  }
  @media screen and (max-width: $break-xs) {
    height: $cell-size-xs;
    width: $cell-size-xs;
  }
  &.active {
    background-color: $dark;
  }
}