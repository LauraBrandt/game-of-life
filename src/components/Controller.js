import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { tick, setRandom, setPattern, clearBoard } from '../actions/boardActions';
import { incrementGeneration, resetGeneration } from '../actions/generationActions';
import '../styles/css/Controller.css';

import patterns from '../startPatterns';

class Controller extends Component {
  constructor() {
    super();

    this.state = {
      timer: null,
      isPlaying: false,
      dropdownOpen: false,
    }

    this.togglePlay = this.togglePlay.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.clear = this.clear.bind(this);
    this.randomize = this.randomize.bind(this);
    this.setPattern = this.setPattern.bind(this);
    this.toggleDropdown = this.toggleDropdown.bind(this);
  }

  togglePlay() {
    if (this.state.isPlaying) {
      clearInterval(this.state.timer);
    } else {
      this.startTimer();
    }
    this.setState( prevState => ({ isPlaying: !prevState.isPlaying }) );
  }

  startTimer() {
    const timer = setInterval(() => {
      this.props.tick();
      this.props.incrementGeneration();
    }, 250);
    this.setState({ timer });
  }

  clear() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.clearBoard();
    this.props.resetGeneration();
  }

  randomize() {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.randomize();
  }

  setPattern(cells) {
    clearInterval(this.state.timer);
    this.setState({ isPlaying: false });
    this.props.resetGeneration();
    this.props.setPattern(cells);
  }

  toggleDropdown() {
    this.setState(prevState => ({dropdownOpen: !prevState.dropdownOpen}));
  }

  componentDidMount() {
    this.randomize();
  }

  componentWillUnmount() {
    clearInterval(this.state.timer);
  }

  render() {
    return (
      <div className="controller">
        <div className="controller-generation">Generation: <span>{this.props.generation}</span></div>
        <hr/>
        <div className="controller-controls">
          <p>Controls</p>
          <button className="controller-button controller-toggle-play" onClick={this.togglePlay}>
            {this.state.isPlaying ? 'Pause' : 'Play'}
          </button>
          <button className="controller-button controller-clear" onClick={this.clear}>
            Clear Board
          </button>
        </div>
        <hr/>
        <div className="controller-start-config">
          <p>Choose a starting configuration (or click the squares on the board to make your own)</p>
          <button className="controller-button controller-random" onClick={this.randomize}>
            Random
          </button>
          <div className="controller-patterns">
            <button className="controller-button controller-pattern-dropdown" onClick={this.toggleDropdown}>
              Patterns &#9662;
            </button>
            {this.state.dropdownOpen &&
              <ul className="controller-pattern-dropdown-list"> 
                {patterns.map(pattern => 
                  <li key={pattern.code} className="controller-dropdown-element">
                    <button className="controller-dropdown-button" onClick={this.setPattern.bind(null, pattern.cells)}>
                      {pattern.name}
                    </button>
                  </li>
                )}
              </ul>
            }
          </div>
        </div>
      </div>
    );
  }
}

Controller.propTypes = {
  generation: PropTypes.number.isRequired,
  tick: PropTypes.func.isRequired,
  randomize: PropTypes.func.isRequired,
  setPattern: PropTypes.func.isRequired,
  clearBoard: PropTypes.func.isRequired,
  incrementGeneration: PropTypes.func.isRequired,
  resetGeneration: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  generation: state.generation
});

const mapDispatchToProps = dispatch => ({
  randomize: () => dispatch(setRandom()),
  setPattern: pattern => dispatch(setPattern(pattern)),
  tick: () => dispatch(tick()),
  clearBoard: () => dispatch(clearBoard()),
  incrementGeneration: () => dispatch(incrementGeneration()),
  resetGeneration: () => dispatch(resetGeneration()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Controller);
