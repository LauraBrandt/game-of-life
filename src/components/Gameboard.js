import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toggleCell } from '../actions/boardActions';
import '../styles/css/Gameboard.css';

export class Gameboard extends Component {
  render() {
    const { board, toggleCell } = this.props;

    return (
      <div className="gameboard">
        { board.map((row, i) => 
          <div className="board-row" key={i}>
            {row.map((cell, j) => 
              <div 
                className={`board-cell${cell ? ' active' : ''}`} 
                key={j} 
                onClick={toggleCell.bind(null, [i, j])}>
              </div>
            )}
          </div>
        )}
      </div>
    );
  }
}

Gameboard.propTypes = {
  board: PropTypes.array.isRequired,
  toggleCell: PropTypes.func
}

const mapStateToProps = state => ({
  board: state.board.current
});

const mapDispatchToProps = dispatch => ({
  toggleCell: cell => dispatch(toggleCell(cell))
});

export default connect(mapStateToProps, mapDispatchToProps)(Gameboard);
