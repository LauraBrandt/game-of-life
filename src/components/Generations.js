import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import '../styles/css/Generations.css';

export class Generations extends Component {
  render() {
    return (
      <div className="generations">
        Generation: <span>{this.props.generation}</span>
      </div>
    );
  }
}

Generations.propTypes = {
  generation: PropTypes.number.isRequired
}

const mapStateToProps = state => ({
  generation: state.generation
});

export default connect(mapStateToProps)(Generations);
