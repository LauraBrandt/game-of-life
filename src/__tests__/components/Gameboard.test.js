import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import { Gameboard } from '../../components/Gameboard';

describe('<Gameboard/>', () => {
  const props = {
    board: [
      [0,1,0],
      [1,0,1]
    ],
    toggleCell: jest.fn()
  }
  let wrapper;

  describe('content', () => {
    beforeEach(() => {
      wrapper = shallow(<Gameboard {...props}/>);
    });

    it('has the correct number of rows', () => {
      expect(wrapper.find('.board-row')).toHaveLength(2);
    });

    it('has the correct number of cells', () => {
      expect(wrapper.find('.board-cell')).toHaveLength(6);
    });

    it('has the correct number of active cells', () => {
      expect(wrapper.find('.active')).toHaveLength(3);
    });
  });

  describe('interactions', () => {
    describe('toggleCell', () => {
      beforeEach(() => {
        wrapper = shallow(<Gameboard {...props}/>);
      });
      afterEach(() => {
        props.toggleCell.mockClear();
      });

      it('calls toggleCell when cell clicked, with target cell as argument', () => {
        const firstCell = wrapper.find('.board-cell').first();
        firstCell.simulate('click');
        expect(props.toggleCell).toBeCalled();
        expect(props.toggleCell).toBeCalledWith([0,0]);
      });
    });
  });

  describe('snapshot', () => {
    it('renders correctly', () => {
      const component = renderer.create(<Gameboard board={[[0,1],[1,0]]} toggleCell={jest.fn()}/>);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});