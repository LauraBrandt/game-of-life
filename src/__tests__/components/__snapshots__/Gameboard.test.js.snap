// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gameboard/> snapshot renders correctly 1`] = `
<div
  className="gameboard"
>
  <div
    className="board-row"
  >
    <div
      className="board-cell"
      onClick={[Function]}
    />
    <div
      className="board-cell active"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <div
      className="board-cell active"
      onClick={[Function]}
    />
    <div
      className="board-cell"
      onClick={[Function]}
    />
  </div>
</div>
`;
